### PR TITLE
Added normalization strategies for numerical types

### PR DESCRIPTION
**Pull request by lordeddard**
# Pull Request

 This PR does these things. 
- Refereces - #296.
- Incorporates two normalization strategies - 'zscore' and 'minmax'.

TODO:
- Will update the user guide after getting some instructions on next steps.

### CODE REVIEW

File: ludwig/features/numerical_feature.py [hunk 1]
```diff
@@ -409,4 +438,4 @@ def populate_defaults(output_feature):
         set_default_value(output_feature, 'clip', None)
         set_default_value(output_feature, 'dependencies', [])
         set_default_value(output_feature, 'reduce_input', SUM)
-        set_default_value(output_feature, 'reduce_dependencies', SUM)
+        set_default_value(output_feature, 'reduce_dependencies', SUM)
```
Comment (msaisumanth) [inline ludwig/features/numerical_feature.py]: please fix @lordeddard

Comment (lordeddard) [reply ludwig/features/numerical_feature.py]: Fixed it! @msaisumanth

File: ludwig/features/numerical_feature.py [hunk 2]
```diff
@@ -47,12 +47,32 @@ def __init__(self, feature):
...
+        'fill_value': 0,
+        'normalization_strategy': None
     }
 
     @staticmethod
     def get_feature_meta(column, preprocessing_parameters):
-        return {}
+        if preprocessing_parameters['normalization_strategy'] is not None:
+            if preprocessing_parameters['normalization_strategy']=='zscore':
+                return {'stats': {
```
Comment (w4nderlust) [inline ludwig/features/numerical_feature.py]: stats is probably redundant, you can put the two values directly in the returned dictionary (here and everywhere else obviously)

**Comment by w4nderlust** [comment]
Thank you really much @lordeddard ! This looks good. i left just a single comment for a minor change. Also please try to respect pip8 in formatting, like space before and ater `-` or `/`. If you use pycharm, `Code > Reformat Code` wold take care of it.
Will merge once you're done with these 2 things.

File: ludwig/features/numerical_feature.py (hunk 2 above)
Comment (lordeddard) [reply ludwig/features/numerical_feature.py]: Changed it!

File: ludwig/features/numerical_feature.py [hunk 3]
```diff
@@ -47,12 +47,32 @@ def __init__(self, feature):
...
     @staticmethod
     def get_feature_meta(column, preprocessing_parameters):
-        return {}
+        if preprocessing_parameters['normalization_strategy'] is not None:
+            if preprocessing_parameters['normalization_strategy'] == 'zscore':
+                return {
+                    'mean': column.astype(np.float32).mean(),
+                    'std': column.astype(np.float32).std()
+                }
+            elif preprocessing_parameters['normalization_strategy'] == 'minmax':
```
Comment (lordeddard) [inline ludwig/features/numerical_feature.py]: Followed PEP8 style guide everywhere except this line. Couldn't get the character count to below 80. @w4nderlust

Comment (msaisumanth) [reply ludwig/features/numerical_feature.py]: you can wrap the elif condition in brackets and split the long line (without using '\')

Comment (w4nderlust) [reply ludwig/features/numerical_feature.py]: actually, even simpler, let's use `normalization` instead of `normalization_strategy` everywhere ;)

Comment (lordeddard) [reply ludwig/features/numerical_feature.py]: Changed the `preprocessing_feature` to `normalization` :)

**Comment by w4nderlust** [comment]
Ok now it's perfect. Now the last two pieces are adding the description of the new parameter in the section `Numerical features preprocessing` in `mkdocs/docs/user_guide.md` and
 adding a test to confirm that this works (@msaisumanth can give you directions on this if you need)

**Comment by lordeddard** [comment]
> adding a test to confirm that this works (@msaisumanth can give you directions on this if you need)

@msaisumanth can you guide me along writing the test cases. Thanks in advance!

**Comment by msaisumanth** [comment]
Oh yes, that's a great idea. It should be simple to add tests. There's a couple of things you can do

1. Integration tests. If you look at:
```
tests/integration_tests/test_experiment.py
```
You'll see a bunch of experiments being run with different model configurations. You can add your normalization schemes there.

2. Unit tests. Create a few numerical feature columns, try both min-max and zero mean normalization and verify the values for correctness. Example tests:
```
tests/ludwig/utils/test_image_utils.py
```

Run the tests with this command in ludwig home directory:
```
python -m pytest
``` 

Let me know if you need more help and Thanks for working on this!

**Comment by lordeddard** [comment]
@w4nderlust : Updated the documentation for the `normalizationn` parameter.
@msaisumanth : Wrote some basic unit-tests and made changes in some integration tests.

**Comment by w4nderlust** [comment]
Looks good to me, wait on @msaisumanth to confirm the test part.

File: tests/ludwig/utils/test_normalization.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    assert feature_1.get_feature_meta(
+        column, {'normalization': 'zscore'}
+    )['mean'] == 6
+    assert feature_2.get_feature_meta(
+        column, {'normalization': 'minmax'}
+    )['min'] == 2
+    assert feature_2.get_feature_meta(
+        column, {'normalization': 'minmax'}
+    )['max'] == 10
+
```
Comment (msaisumanth) [inline tests/ludwig/utils/test_normalization.py]: This is great. But could you please add a couple of checks validating the normalized values?

**Comment by msaisumanth** [comment]
Added a suggestion

File: tests/ludwig/utils/test_normalization.py (hunk 4 above)
Comment (msaisumanth) [reply tests/ludwig/utils/test_normalization.py]: e.g. after zscore normalization, the first row should be close to -1.4144. You can use np.allclose to verify

Comment (lordeddard) [reply tests/ludwig/utils/test_normalization.py]: Added the tests. Used [Unbiased Estimator](https://stackoverflow.com/questions/24984178/different-std-in-pandas-vs-numpy)(Pandas) for standard deviation calculation.

Comment (lordeddard) [reply tests/ludwig/utils/test_normalization.py]: @msaisumanth , Is there anything more I have to add to the `tests`?

Comment (msaisumanth) [reply tests/ludwig/utils/test_normalization.py]: looks good!

**Comment by msaisumanth** [comment]
Looks good! Thanks @lordeddard 

 @w4nderlust

**Comment by w4nderlust** [comment]
Thank you really much @lordeddard !

**Pull request merged**